search/search_quality/merge_samples.py

```diff
@@ -0,0 +1,181 @@
+import argparse
+import copy
+import json
+
+
+def read_samples(path):
+    with open(path, "r") as f:
+        return [json.loads(line) for line in f]
+
+
+def doubles_equal(a, b):
+    EPS = 1e-6
+    return abs(a - b) < EPS
+
+
+def points_equal(a, b):
+    return doubles_equal(a["x"], b["x"]) and doubles_equal(a["y"], b["y"])
+
+
+def rects_equal(a, b):
+    for key in ["minx", "miny", "maxx", "maxy"]:
+        if not doubles_equal(a[key], b[key]):
+            return False
+    return True
+
+
+def samples_equal(a, b):
+    """
+    Returns whether two samples originate from the same unassessed sample,
+    i.e. found results and uselessness are not taken into account.
+    """
+    for field in ["query", "locale"]:
+        if a[field] != b[field]:
+            return False
+
+    pos_key = "position"
+    bad_point = {"x": -1, "y": -1}
+    a_pos = a[pos_key] or bad_point
+    b_pos = b[pos_key] or bad_point
+    if not points_equal(a_pos, b_pos):
+        return False
+
+    vp_key = "viewport"
+    if not rects_equal(a[vp_key], b[vp_key]):
+        return False
+
+    return True
+
+
+def results_equal(a, b):
+    """
+    Returns whether two found results are equal before assessing.
+    Does not take relevance into account.
+    """
+    pos_key = "position"
+    name_key = "name"
+    hn_key = "houseNumber"
+    types_key = "types"
+    if a[name_key] != b[name_key]:
+        return False
+    if not points_equal(a[pos_key], b[pos_key]):
+        return False
+    if a[hn_key] != b[hn_key]:
+        return False
+    if sorted(a[types_key]) != sorted(b[types_key]):
+        return False
+    return True
+
+
+def greedily_match_results(a, b):
+    match_to_a = [-1] * len(a)
+    match_to_b = [-1] * len(b)
+    for i in range(len(a)):
+        if match_to_a[i] >= 0:
+            continue
+        for j in range(len(b)):
+            if match_to_b[j] >= 0:
+                continue
+            if results_equal(a[i], b[j]):
+                match_to_a[i] = j
+                match_to_b[j] = i
+                break
+    return match_to_a, match_to_b
+
+
+def merge_relevancies(a, b):
```

Comment:
doc string at least on the protocol will be much appreciated

Reply:
done